porch/repository/pkg/git/draft.go

```diff
@@ -75,84 +69,73 @@ func (d *gitPackageDraft) UpdateLifecycle(ctx context.Context, new v1alpha1.Pack
-		case nil:
-			// Tag exists, cannot overwrite
-			return nil, fmt.Errorf("another instance of finalized package %s:%s:%s already exists", d.parent.name, d.path, d.revision)
-		case plumbing.ErrReferenceNotFound:
-			// Tag does not yet exist
-		}
-
-		newRef, err := setReference(repo.Storer, refNames.final, commitHash)
-		if err != nil {
-			return nil, fmt.Errorf("failed to finalize package: %v", err)
-		}
+		tag := createFinalTagNameInLocal(d.path, d.revision)
+		pushHelper.Push(commitHash, r.branch.RefInLocal()) // Push new main branch
+		pushHelper.Push(commitHash, tag)                   // Push the tag
+		pushHelper.RequireRef(commitBase)                  // Make sure main didn't advance
 
-		pushRefSpecs = append(
-			pushRefSpecs,
-			config.RefSpec(fmt.Sprintf("%s:%s", commitHash, remoteRef)),      // Push new main
-			config.RefSpec(fmt.Sprintf("%s:%s", commitHash, refNames.final)), // Push the tag
-		)
-
-		if d.ref != nil {
-			cleanup[d.ref.Name()] = true
+		// Delete base branch?
+		switch base := d.base; {
+		case base == nil: // no branch to delete
+		case base.Name() == draftBranch.RefInLocal(), base.Name() == proposedBranch.RefInLocal():
+			pushHelper.Delete(base)
 		}
-		// also make sure to clean up draft and proposed refs for this package, if they exist.
-		cleanup[refNames.draft] = true
-		cleanup[refNames.proposed] = true
 
 		// Update package draft
 		d.commit = commitHash
 		d.tree = newTreeHash
-		d.ref = newRef
+		newRef = plumbing.NewHashReference(tag, commitHash)
 
 	case v1alpha1.PackageRevisionLifecycleProposed:
-		newRef, err := setReference(repo.Storer, refNames.proposed, d.commit)
-		if err != nil {
-			return nil, fmt.Errorf("failed to update package lifecycle to \"proposed\": %v", err)
-		}
 		// Push the package revision into a proposed branch.
-		pushRefSpecs = append(pushRefSpecs, config.RefSpec(fmt.Sprintf("%s:%s", d.commit, refNames.proposed)))
-		cleanup[refNames.draft] = true
+		pushHelper.Push(d.commit, proposedBranch.RefInLocal())
+
+		// Delete base branch?
+		switch base := d.base; {
+		case base == nil: // no branch to delete
+		case base.Name() != proposedBranch.RefInLocal():
+			pushHelper.Delete(base)
+		}
 
 		// Update package referemce (commit and tree hash stay the same)
-		d.ref = newRef
+		newRef = plumbing.NewHashReference(proposedBranch.RefInLocal(), d.commit)
 
 	case v1alpha1.PackageRevisionLifecycleDraft:
-		newRef, err := setReference(repo.Storer, refNames.draft, d.commit)
-		if err != nil {
-			return nil, fmt.Errorf("failed to update package draft: %v", err)
-		}
 		// Push the package revision into a draft branch.
-		// TODO: implement config resolution rather than forcing the push (+)
-		pushRefSpecs = append(pushRefSpecs, config.RefSpec(fmt.Sprintf("+%s:%s", d.commit, refNames.draft)))
-		cleanup[refNames.proposed] = true // In case client downgraded packaget to draft from proposed
+		pushHelper.Push(d.commit, draftBranch.RefInLocal())
+		// Delete base branch?
```

Comment:
Maybe instead of "Delete base branch?" (which looks a bit like a TODO), we should say "Delete base branch (if exists)"

Reply:
done. thanks!